### PR TITLE
[TRANSFORMATIONS] Add Softmax sinking throught TopK into CommonOptimizations

### PR DESCRIPTION
**Pull request by EgorDuplensky**
### Details:
 - the use case exists in one of moe routing patterns

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+TopkRenormalizeToSoftmaxAfterTopkFusion::TopkRenormalizeToSoftmaxAfterTopkFusion() {
+    MATCHER_SCOPE(TopkRenormalizeToSoftmaxAfterTopkFusion);
+
+    auto p_logits = pattern::any_input();
+    // Softmax feeds only this TopK; else dropping it changes other consumers.
+    auto p_softmax = pattern::wrap_type<v1::Softmax, v8::Softmax>({p_logits}, pattern::consumers_count(1));
+    auto p_topk_k = pattern::any_input();
+    // TopK.values (port 0) feeds only ReduceSum and Divide; extras would observe softmax-probs replaced by raw logits.
+    auto p_topk = pattern::wrap_type<v1::TopK, v3::TopK, v11::TopK>({p_softmax, p_topk_k}, pattern::consumers_count(2));
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]:
```suggestion
    auto p_topk = pattern::wrap_type<ov::op::util::TopKBase>({p_softmax, p_topk_k}, pattern::consumers_count(2));
```

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+        // Recover Softmax axis.
+        int64_t softmax_axis = 0;
+        if (auto sm8 = ov::as_type_ptr<v8::Softmax>(softmax)) {
+            softmax_axis = sm8->get_axis();
+        } else if (auto sm1 = ov::as_type_ptr<v1::Softmax>(softmax)) {
+            softmax_axis = static_cast<int64_t>(sm1->get_axis());
+        } else {
+            return false;
+        }
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: @mitruska should we create a parent class for Softmax one day?

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        // TopKBase unifies get_axis() across v1/v3/v11.
+        const auto topk_axis = static_cast<int64_t>(ov::as_type_ptr<ov::op::util::TopKBase>(topk)->get_axis());
+
+        // The renormalize must operate on the same axis as the TopK selection.
+        const auto rank = divide->get_output_partial_shape(0).rank().get_length();
+        auto norm_axis = [rank](int64_t a) {
+            return a >= 0 ? a : a + rank;
+        };
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: I'm pretty sure, there're some rank normalization utils in OpenVINO already.

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        auto red_axes_const = ov::as_type_ptr<v0::Constant>(pm.at(p_red_axes).get_node_shared_ptr());
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp:65]: I know, it might be irritating, but since you're casting, please check the pointer first
```suggestion
        auto red_axes_const = ov::as_type_ptr<v0::Constant>(pm.at(p_red_axes).get_node_shared_ptr());
        if (!red_axes_const) {
            return false;
        }
```

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        new_topk->set_friendly_name(topk->get_friendly_name());
+
+        // Build a new SoftMax over new TopK.values; this reproduces the
+        // routing-weight tensor that the original Divide produced.
+        std::shared_ptr<ov::Node> new_softmax;
+        if (ov::as_type_ptr<v8::Softmax>(softmax)) {
+            new_softmax = std::make_shared<v8::Softmax>(new_topk->output(0), softmax_axis);
+        } else {
+            new_softmax = std::make_shared<v1::Softmax>(new_topk->output(0), static_cast<size_t>(softmax_axis));
+        }
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: I'm not sure there's any point having this conditional statement. @mitruska can we just create `v8::Softmax` and get away with it?

**Comment by CuriousPanCake** [comment]
I think, the title should be changed to [TRANSFORMATIONS] since you're adding a new transformation :)

**Comment by CuriousPanCake** [comment]
Could you please add unit tests for the new transformation?

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp (hunk 2 above)
Comment (mitruska) [reply src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: As it can simplify such pipelines and pattern detection - it could be done, but at the same time we can consider deprecation of the v1::Softmax.

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp (hunk 5 above)
Comment (mitruska) [reply src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: It depends on what is observed in the target model IR, but as `ConvertSoftMax8ToSoftMax1` is still a part of common optimizations it's better to keep it compatible with both versions. https://github.com/openvinotoolkit/openvino/blob/c9abcacdc472eb4fe76deb3903469ad1bf86674e/src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp#L222

**Review by CuriousPanCake — CHANGES_REQUESTED** [review]

**Comment by EgorDuplensky** [comment]
@CuriousPanCake Thank you for the comments. Applied

File: src/common/transformations/tests/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+#include "openvino/op/parameter.hpp"
+#include "openvino/op/reduce_sum.hpp"
+#include "openvino/op/result.hpp"
+#include "openvino/op/softmax.hpp"
+#include "openvino/op/topk.hpp"
+
+using namespace testing;
+using namespace ov;
+
+namespace {
```
Comment (praasz) [inline src/common/transformations/tests/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: Remove it and wrap all test in:
`namespace ov::test {`

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    auto p_logits = pattern::any_input();
+    // Softmax feeds only this TopK; else dropping it changes other consumers.
+    auto p_softmax = pattern::wrap_type<v1::Softmax, v8::Softmax>({p_logits}, pattern::consumers_count(1));
+    auto p_topk_k = pattern::any_input();
+    // TopK.values (port 0) feeds only ReduceSum and Divide; extras would observe softmax-probs replaced by raw logits.
+    auto p_topk = pattern::wrap_type<ov::op::util::TopKBase>({p_softmax, p_topk_k}, pattern::consumers_count(2));
+    auto p_red_axes = pattern::wrap_type<v0::Constant>();
+    auto p_reduce =
+        pattern::wrap_type<v1::ReduceSum>({p_topk, p_red_axes}, pattern::attrs_match({{"keep_dims", true}}));
+    auto p_divide = pattern::wrap_type<v1::Divide>({p_topk, p_reduce}, pattern::has_static_rank());
```
Comment (mryzhov) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp:40]: ideally to check that p_topk is on port 0 and p_reduce is on port 1, but don't think that is a real case

**Review by mryzhov — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+TopkRenormalizeToSoftmaxAfterTopkFusion::TopkRenormalizeToSoftmaxAfterTopkFusion() {
+    MATCHER_SCOPE(TopkRenormalizeToSoftmaxAfterTopkFusion);
+
+    auto p_logits = pattern::any_input();
+    // Softmax feeds only this TopK; else dropping it changes other consumers.
+    auto p_softmax = pattern::wrap_type<v1::Softmax, v8::Softmax>({p_logits}, pattern::consumers_count(1));
+    auto p_topk_k = pattern::any_input();
+    // TopK.values (port 0) feeds only ReduceSum and Divide; extras would observe softmax-probs replaced by raw logits.
+    auto p_topk = pattern::wrap_type<ov::op::util::TopKBase>({p_softmax, p_topk_k}, pattern::consumers_count(2));
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp]: Use predicate `pattern::output_index_matches(0)` instead of the check in the callback below.

File: src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        auto topk = pm.at(p_topk).get_node_shared_ptr();
+        auto reduce = pm.at(p_reduce).get_node_shared_ptr();
+        auto divide = pm.at(p_divide).get_node_shared_ptr();
+
+        // The renormalization must consume TopK.values (port 0), not indices
+        if (pm.at(p_topk).get_index() != 0) {
+            return false;
+        }
+
+        const auto rank = divide->get_output_partial_shape(0).rank();
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/topk_renormalize_to_softmax_after_topk_fusion.cpp:50]: You've already checked it in the pattern

**Comment by CuriousPanCake** [comment]
@EgorDuplensky is this one still relevant?
